torch/csrc/dynamo/compiled_autograd.h

```diff
@@ -615,6 +674,36 @@ class SwapSavedVariables {
     }
   }
 
+  template <typename K, typename V>
+  void before(ska::flat_hash_map<K, V>& m) {
+    for (auto& [k, v] : m) {
+      before(k);
+      before(v);
+    }
+  }
+
+  template <typename K, typename V>
+  void after(ska::flat_hash_map<K, V>& m) {
+    for (auto& [k, v] : m) {
+      before(k);
+      before(v);
+    }
+  }
+
+  template <typename T>
+  void before(std::unordered_set<T>& s) {
+    for (auto& i : s) {
+      before(i);
+    }
+  }
+
+  template <typename T>
+  void after(std::unordered_set<T>& s) {
+    for (auto& i : s) {
+      after(i);
+    }
+  }
```

Comment:
I believe we depend on the order of iteration matching between collect() and before() in some cases.

Reply:
hmm this is not obvious to me, do you have an example? I can add a test for it, and sort or restrict types 

my understanding:
- collect() interacts with the cache key and lifted args
- before() interacts with the stash and lifted args 
- lifted args only has 3 ordered containers: all_size_inputs, dyn_size_inputs and hooks
- but we don't call before() on any of them, so there should be no ordering conflict

Reply:
If we create dynamic shapes (because an SymInt changes).

Reply:
i added a sort in case

right now for tensors saved in saved_data (ska::flat_hash_map<std::string, c10::IValue>), we specialize on the memory addresses of tensors impl (which are different even for views).